rfcs/theme-shape.md

```diff
@@ -0,0 +1,228 @@
+
+_(This change has no direct impact on perf, but is required for the [next point](#flatten-tokens).)_
+
+To be able to do this, we need get rid of any naming overlaps between alias and global tokens.
+The only potential overlap is in spacing tokens as we have both global and alias spacing tokens -> talk to design to understand which tokens engineers are supposed to use.
+
+Current design token spec also calls out `fontWeights` and `alignment` tokens are "not actual tokens" and should not be used -> talk to design, should we remove them?
+
+### <a name="flatten-tokens"></a>Flatten the tokens object
+
+In Typescript the theme is represented by a deep object:
+
+```js
+/* Before */
+theme.alias.colors.red.neutral;
+```
+
+This brings the three following issues:
+
+1. Engineers might struggle to find the expected token in the object.
+2. Theme merging is expensive as we need to merge deep objects.
+3. Insertion of CSS variables is more expensive as we need to flatten CSS variables in runtime (`flattenThemeToCSSVariables()`).
+
+**The proposal is to merge all the tokens to a single flat object**:
+
+```js
+/* After */
+theme.aliasColorsRedNeutral;
+```
+
+This should simplify theme merging, improve tokens discoverability. The potential drawback is we are creating a gigantic dictionary of unrelated values.
+
+### <a name="hash-variables"></a>Hash variable names
+
+As described in [Performance analysis](#performance-analysis), total string length of CSS variable names is ~40KB, by hashing the variable names to 7 char hashes, we can reduce this string length by 75%.
+
+The performance analysis shows no perf impact (in rendering phase) by doing so. Also when using the variables in `makeStyles`, we need either hash the names on the fly (perf hit) or reference a dictionary (potential bundle size and memory hit).
+
+Besides the performance, another valid reason for hashing the variable names might be **encapsulation**. Currently, the CSS variable names are deterministic, anybody can use them directly in both CSSInJS and SASS/CSS - this would make the variable names part of the API contract and any variable name change would be a breaking change. Hashing the variable names can prevent this misuse.
+
+**Let's create a separate RFC and have a discussion on this topic after beta.**
+
+### <a name="be-cautious"></a>Be cautious when adding new tokens to the theme
+
+There are currently ~1200 tokens in the theme object. ~1000 of the tokens represent global and alias shared colors which are actually rarely used in products.
+80% of tokens rarely used.
+
+**When adding tokens, think about the number of CSS variables added, whether they change during runtime and need to be implemented using CSS variables.**
+
+### <a name="no-css-vars"></a>Get rid of all CSS variables
+
+We have learnt that CSS variables are expensive. Can we get rid of them completely? If we decided to do so, we would need to think about the following problems:
```

Comment:
Can we quantify what expensive means and what too expensive is?  Devs integrating other controls may be familiar with and prefer CSS vars.

Reply:
This is the best I have for now: https://github.com/microsoft/fluentui/blob/bc239d778215e46814b78b656a97345fef8d5a29/rfcs/theme-shape.md#results-summary

It shows that by having more CSS variables, the performance of the synthetic perf example degrades. I have to admit it is too vague, especially with the y axis missing any units.

It would be great to do deeper investigation as described here: https://github.com/microsoft/fluentui/pull/19752#discussion_r709221136